config.org
#+STARTUP: overview
#+OPTIONS: toc:2

* OPTIONS
:PROPERTIES:
:header-args: :tangle  ~/.config/nvim/lua/config/options.lua
:END:

** TODO Task list
- [ ] automatically refile those options?
- [ ] find nice configs/categories
- [ ] function keys also useful so u can use them in insert/visual mode whichkey mappings

#+begin_src lua
-- let mapleader = "\<Space>"
-- vim.g.mapleader = " "
vim.g.mapleader = vim.keycode"<space>"
-- for languages/filetype mappings?
-- vim.g.maplocalleader = vim.keycode[[<f15>]]
vim.g.maplocalleader = vim.keycode[[\]]
-- not sure
-- vim.g.maplocalleader = vim.keycode[[<c-c>]]

-- Tip: per tradurre più parole usa visual mode
-- autocmd Filetype markdown,vimwiki set keywordprg=trans\ :it
-- autocmd Filetype markdown,vimwiki set keywordprg=trans\ en:it

vim.g.health = { style = 'float' }

-- vim.o.winborder='border'

-- hi Normal guibg=NONE ctermbg=NONE

-- hi Cursor guifg=green guibg=green
-- TODO: hide cursor after some time?
vim.cmd[[
" change color when inside colored text? like hex plugin?
" light green when dark background, rose when light background
hi Cursor2 guifg=red guibg=#00ffa0
"maybe use in insert mode?
set guicursor=n-v-c:block-Cursor/lCursor,i-n:Cursor2/lCursor2
]]

-- set guicursor=n-c-v-i:block-nCursor
-- set guicursor=i:blinkoff400-blinkon250

-- set guifont=DroidSansMono\ Nerd\ Font\ 11

vim.opt.inccommand="split"
#+end_src

Categories from =:options=:

** Important
#+begin_src lua
#+end_src

** Moving around, searching and patterns
#+begin_src lua
vim.opt.ignorecase = true       -- Ignore case letters when search
vim.opt.smartcase = true        -- Ignore lowercase for the whole pattern
vim.opt.path:append('**') -- make the search recursive
vim.opt.startofline = false -- Keep same column when going visual-block column
-- don't use included files and tag complete for better performances. Add this to your vimrc for these options :
-- go to previous/next line with h,l,left arrow and right arrow when cursor reaches end/beginning of line
-- opt.whichwrap:append "<>[]hl"
#+end_src

** Tags
#+begin_src lua
#+end_src

** Displaying text
#+begin_src lua
vim.opt.showbreak="↪"
-- how to show full parent directory when the window is small?
-- vim.opt.undoreload = 10000        -- number of lines to save for undo
vim.opt.wrap = false
vim.opt.list = false
vim.opt.listchars = { trail = "-", tab = [[▸ ]], nbsp = "␣", eol = "$", extends = "❯", precedes ="❮" }
-- vim.opt.fillchars:append { diff = "╱" }
-- vim.opt_global.fillchars:append { diff = "╱" }
-- opt.fillchars = { eob = " " }
-- vim.opt.lazyredraw = true       -- Faster scrolling
vim.opt.number = true           -- Show line number
-- vim.opt.colorcolumn = '80'      -- Line lenght marker at 80 columns
vim.opt.linebreak = true        -- Wrap on word boundary
-- Enable relative numbers only in Normal mode, and absolute numbers only in Insert mode
-- augroup toggle_relative_number
-- autocmd InsertEnter * :setlocal norelativenumber nonumber
-- autocmd InsertLeave * :setlocal relativenumber number
-- display completion matches on your status line
#+end_src

** Syntax, highlighting and spelling
#+begin_src lua
-- vim.wo.spell = true
-- vim.o.spelllang    = 'en,uk,it,ru,ar'   -- Define spelling dictionaries
vim.o.spelloptions = 'camel'      -- Treat parts of camelCase words as seprate words
vim.opt.synmaxcol = 240         -- Max column for syntax highlight
vim.opt.cursorline = true -- Highlight current line
-- vim.opt.colorcolumn = '80'      -- Line lenght marker at 80 columns
vim.opt.termguicolors = false    -- Enable 24-bit RGB colors
-- https://github.com/dettonijr/spass-syntax-highlight
-- au BufRead,BufNewFile *.dfg set filetype=dfg

vim.cmd('hi Visual guibg=peru guifg=Black')
-- hi Visual guibg=#000066 guifg=#00ff99
vim.api.nvim_set_hl(0, 'Pmenu', { bg = 'NONE' })
vim.api.nvim_set_hl(0, 'PmenuBorder', { bg = 'NONE', fg = "#CC6600" })
#+end_src

** Multiple windows
#+begin_src lua
-- exclude hidden buffers when saving session {for possession.nvim}
vim.o.switchbuf    = 'usetab'       -- Use already opened buffers when switching
vim.opt.hidden = true           -- Enable background buffers
-- like emacs
vim.opt.splitright = true       -- Vertical split to the right
vim.opt.splitbelow = true       -- Horizontal split to the bottom
-- vim.opt.laststatus = 2
vim.opt.laststatus=3            -- Set global statusline
-- vim.cmd[[let &stc='%=%{v:virtnum>0?"":printf("%X",v:lnum)} ']]
#+end_src

** Multiple tab pages
#+begin_src lua
-- vim.opt.showtabline = 1
-- set showtabline=2
#+end_src

** Terminal
#+begin_src lua
vim.o.title = true
#+end_src

** Using the mouse
#+begin_src lua
-- doesn't work w/ noice command line
vim.opt.mouse = 'a'
vim.opt.mousescroll  = 'ver:25,hor:6' -- Customize mouse scroll
#+end_src

** Messages and info
#+begin_src lua
vim.opt.showmode = false -- Not show redundant mode line with airline
-- vim.opt.more = false
vim.opt.shortmess:append "sI" -- Disable nvim intro
-- opt.shortmess:append('filmnrxoOtT') -- Show shorten messages in UI
#+end_src

** Selecting text
#+begin_src lua
#+end_src

** Editing text
#+begin_src lua
vim.opt.joinspaces = false
vim.opt.complete:remove('t')      -- Don't use tags for completion
vim.opt.complete:append('kspell') -- Add spellcheck options for autocomplete
vim.opt.dictionary = '/usr/share/dict/words'
vim.opt.undolevels = 10000         -- use many levels of undo
--   set undodir=$HOME/.vim/undo
--   set undofile 
-- vim.opt.undofile = true                -- Save undo's after file closes
-- vim.opt.undodir="$HOME/.vim/undo" -- where to save undo histories
-- vim.opt.undolevels = 1000         -- How many undos
-- vim.opt.undoreload = 10000        -- number of lines to save for undo
vim.opt.showmatch = true        -- Highlight matching parenthesis
vim.opt.matchtime = 1
vim.opt.infercase = true                       -- Adjust completions to match case
-- silent! setglobal dictionary+=/usr/share/dict/words
vim.opt.matchpairs:append('<:>')
vim.opt.formatoptions:remove{ "c", "r", "o" } -- stop auto-insertion of comment symbols.
-- vim.o.formatoptions = 'rqnl1j' -- Improve comment editing
vim.o.formatlistpat = [[^\s*[0-9\-\+\*]\+[\.\)]*\s\+]]
-- [The linematch diffopt makes builtin diff so sweat! : r/neovim](https://www.reddit.com/r/neovim/comments/1ihpvaf/the_linematch_diffopt_makes_builtin_diff_so_sweat/)
-- don't use included files and tag complete for better performances. Add this to your vimrc for these options :
-- set complete-=t
-- set complete-=i

vim.o.pumborder = 'rounded'
#+end_src

** Tabs and indenting
#+begin_src lua
vim.o.autoindent = true
vim.opt.expandtab = true       -- When this option is enabled, vi will use spaces instead of tabs
vim.opt.shiftwidth  = 2 -- Determines the amount of whitespace to add in normal mode
vim.opt.smartindent = true
vim.opt.tabstop     = 2 -- Width of tab character
vim.opt.softtabstop = 2 -- Fine tunes the amount of white space to be added
#+end_src

** Folding
#+begin_src lua
-- vim.wo.foldtext = [[substitute(getline(v:foldstart),'\\t',repeat('\ ',&tabstop),'g').'...'.trim(getline(v:foldend)) ]]
-- Sexy Folds (https://www.reddit.com/r/neovim/comments/psl8rq/sexy_folds/)
-- https://www.reddit.com/r/neovim/comments/1h34lr4/neovim_now_has_the_builtin_lsp_folding_support/
vim.wo.foldenable = false
vim.wo.foldlevel = 99
vim.opt.foldlevelstart = 99
-- fallback to treesitter if no lsp capabilities?
-- vim.o.foldenable= false
-- vim.wo.foldtext = [[substitute(getline(v:foldstart),'\\t',repeat('\ ',&tabstop),'g').'...'.trim(getline(v:foldend)) ]]
vim.wo.foldnestmax = 3
vim.wo.foldminlines = 1
-- vim.opt.foldopen = 'block,hor,mark,percent,quickfix,tag,jump,search,undo' -- What movements open folds
#+end_src

** Diff mode
#+begin_src lua
-- vim.opt.fillchars:append { diff = "╱" }
-- vim.opt_global.fillchars:append { diff = "╱" }
-- vim.opt.diffopt:append({'vertical'})
-- [The linematch diffopt makes builtin diff so sweat! : r/neovim](https://www.reddit.com/r/neovim/comments/1ihpvaf/the_linematch_diffopt_makes_builtin_diff_so_sweat/)
vim.opt.diffopt="filler,internal,closeoff,algorithm:histogram,context:5,linematch:60"
-- set diffopt=vertical,filler,internal,algorithm:histogram,indent-heuristic
#+end_src

** Mapping
#+begin_src lua
vim.opt.timeoutlen = 500 -- Mapping waiting time
#+end_src

** Reading and writing files
#+begin_src lua
-- vim.opt.backup = false
-- vim.opt.writebackup = false
-- Accept modeline of each file
vim.opt.modeline = true
vim.opt.modelineexpr = true -- no idea why I can make this local...
vim.opt.modelines = 3 -- after eventual shebang and emacs...
-- opt.fileformat = 'unix' -- End of file setting
#+end_src

** The swap file
#+begin_src lua
-- how to show full parent directory when the window is small?
-- vim.opt.swapfile = false
vim.opt.updatetime = 250        -- ms to wait for trigger an event
#+end_src

** Command line editing
#+begin_src lua
vim.opt.history = 10000           -- Remember N lines in history
-- vim.opt.wildmode = {list = "longest", "full"} -- When completing, show all options, insert common prefix, then iterate
vim.opt.wildignore = 'deps,.svn,CVS,.git,.hg,*.o,*.a,*.class,*.mo,*.la,*.so,*.obj,*.swp,*.jpg,*.png,*.xpm,*.gif,.DS_Store,*.aux,*.out,*.toc'
-- ignore builtin colorschemes for Snacks.picker.colorschemes()
vim.opt.wildignore:append(vim.api.nvim_get_runtime_file("colors/*.{vim,lua}", true))
vim.opt.wildmenu = true
#+end_src

** Executing external commands
#+begin_src lua
vim.o.sh='/usr/bin/env bash'
#+end_src

** Running make and jumping to errors (quickfix)
#+begin_src lua
#+end_src

** Language specific
#+begin_src lua
#+end_src

** Multi-byte characters
#+begin_src lua
-- opt.fileencoding = 'utf-8'
vim.opt.encoding = "utf-8"
#+end_src

** Various
#+begin_src lua
vim.g.sessionoptions = "buffers,curdir,folds,help,tabpages,winsize,terminal" -- removed blank
-- vim.opt.sessionoptions:remove('buffers')
vim.wo.signcolumn = "yes:1"
vim.o.exrc=true
-- aumenta fzf history
vim.opt.shada={"!","'500","<50","s10","h"}
#+end_src

* KEYMAPS
:PROPERTIES:
:header-args: :tangle  ~/.config/nvim/lua/config/keymaps.lua
:END:

** Conventions
- Never use literal spaces in a mapping, use <space> (formatting may take those away and it's more explicit):
  - be explicit with final space: ={ '<leader>gC', ':Git commit -v -q<space>' }=
- Don't use leader(s) mappings for operators (ex: surround), use them for general plugin mappings that don't define an operator (or text object of course)
	- Capitalize desc field: rg 'desc\s*=\s*[[:punct:]]+[a-z]' --vimgrep |v -q-
- Take inspiration from emacs/nano mappings 
- Use Emacs mapping notation (<C-c> and not <c-c> and use <M- instead of <A-)
- Use <f19> or whatever for universal argument (hold u)

** Task list

- [ ] Translation: ~ -> synonym (Google):-nnoremap <leader>~ thesanosaurus :)) (google trans command line synonym)

#+begin_src lua
-- find better way for comment string of multiple chars
-- vim: foldmethod=expr foldexpr=getline(v\:lnum)=~'^--\ ┣'?'>1'\:getline(v\:lnum)=~'^--\ ─'?'>2'\:'=':

-- make these work with dot repeat (repeatable)
-- vim.keymap.set({ 'x', 'n' }, '<localleader>j', "o<esc>p", { buffer = ev.buf, silent = true })
-- vim.keymap.set({ 'x', 'n' }, '<localleader>k', "O<esc>p", { buffer = ev.buf, silent = true })
-- vim.keymap.set('n', '<leader>o', 'o<esc>o', { buffer = ev.buf, desc = 'Add newline below and start insert', nowait=true })
-- maybe make these mappings global?
-- vim.keymap.set('n', '<leader>o', function() if vim.fn.getline('.'):match('^$') then return '<esc>o' else return
-- 'o<esc>o'end end, { buffer = ev.buf, desc = 'Add newline below and start insert', nowait=true, expr = true })
-- vim.keymap.set('n', '<leader>O', 'O<esc>O', { buffer = ev.buf, desc = 'Add newline above and start insert', nowait=true })
-- vim.keymap.set('n', '<leader>O', function() if vim.fn.getline('.'):match('^$') then return '<esc>O' else return
-- 'O<esc>O'end end, { buffer = ev.buf, desc = 'Add newline above and start insert', nowait=true, expr = true })
-- vim.keymap.set('n', '<leader>k', 'O<esc>Vp', { buffer = ev.buf, desc = 'Paste above', nowait=true })
-- vim.keymap.set('n', '<leader>j', 'o<esc>Vp', { buffer = ev.buf, desc = 'Paste below', nowait=true })
-- vim.keymap.set('n', '<leader>K', '2O<esc>Vp', { buffer = ev.buf, desc = 'Paste above', nowait=true })
-- vim.keymap.set('n', '<leader>J', '2o<esc>Vp', { buffer = ev.buf, desc = 'Paste below', nowait=true })



---see https://github.com/nvim-mini/mini.nvim/blob/91f7a680fe5765b68456ab6e7b910d478da083b9/lua/mini/basics.lua#L548-L549

-- https://gist.github.com/kawarimidoll/496cb16b40af33e8d84daff6dde8a16f
-- maybe use 'a' like in :h 'mouse'
-- local all = vim.fn.split('nvsxoilct', [[.\zs]])
local all = vim.fn.split('nvsxoilt', [[.\zs]])
-- modes = { "n", "v", "x", "s", "o", "i", "c", "t" },

 -- scriptease mapping? fallback to that/syntax if no treesitter...
 -- works with right click too
vim.keymap.set('n', 'zS', '<cmd>Inspect<CR>')
vim.keymap.set('n', '<space>[', '<cmd>TermHl<CR>', { desc = 'Colorize the current buffer' })

---see [How to insert newline without entering insert mode? : r/neovim](https://www.reddit.com/r/neovim/comments/10kah18/how_to_insert_newline_without_entering_insert_mode/)
-- Add empty lines before and after cursor line
-- vim.keymap.set('n', '<c-s-cr>', "<Cmd>call append(line('.') - 1, repeat([''], v:count1))<CR>")
-- vim.keymap.set('n', '<c-cr>', "<Cmd>call append(line('.'),     repeat([''], v:count1))<CR>")

-- add mapping in visual mode for commenting and adding todo, etc... (using v:count and gc), like 4gc for HACK or something

-- vim.keymap.set('n','<leader>N','<cmd>tabnew|e /tmp/notes.anki|set ft=anki<cr>')
-- vim.cmd[[
-- " I'd like to know if there's not a window above, if not use <c-w>j
-- nnoremap <expr> k v:count1>=line('.')?'<c-w>k':'k'
-- " I'd like to know if there's not a window below, if not use <c-w>k
-- nnoremap <expr> j v:count1>line('$')-line('.')?'<c-w>j':'j'
-- nnoremap <expr> h v:count1>=charcol('.')?'<c-w>h':'h'
-- nnoremap <expr> l v:count1>charcol('$')-1-charcol('.')?'<c-w>l':'l'
-- ]]

-- /usr/local/share/nvim/runtime/doc/cmdline.txt
-- how to never make it quit? like a repl
vim.cmd([[autocmd CmdwinEnter * map <buffer> <C-CR> <CR>q:]])

-- vim-for-php-programmers.pdf
-- map CTRL-L to piece-wise copying of the line above the current one
-- vim.keymap.set('i','<C-L>', '@@@<ESC>hhkywjl?@@@<CR>P/@@@<CR>3s')
-- fixa per copiare word?
-- these two show up in which-key, FIX
-- vim.keymap.set('i','yW', '@@@<ESC>hhkywjl?@@@<CR>P/@@@<CR>3s')
-- vim.keymap.set('i','yw', '@@@<ESC>hhjywkl?@@@<CR>P/@@@<CR>3s')

-- paste but leave empty lines above/below
-- vim.keymap.set('n', "<KEY>",[[o<esc>p]])
-- vim.keymap.set('n', "<KEY>",[[O<esc>P]])

-- vim.keymap.set('n', 'z<C-g>', "<cmd>call setreg('+', getreg('%').'#'.line('.'))<CR><C-g>")

-- vim.keymap.set('c', '<c-s>', "submatch")
-- check if \= before though
-- vim.keymap.set('c', '\d (da 0 a 9)', "submatch (da0 a 9)")

-- kinda cool: mappings similar to the one with <++> as placeholder
-- vim.keymap.set('n', '<space><space>', "<CMD>bmod<CR>")

--select whatever's just been pasted, or read into the buffer via :r! etc, respecting line/char visual mode. (https://www.reddit.com/r/vim/comments/4aab93/weekly_vim_tips_and_tricks_thread_1/)
-- vim.keymap.set('n', 'gV', function() return '`[' .. vim.fn.strpart(vim.fn.getregtype(), 0, 1) .. '`]' end, { expr = true })
vim.keymap.set('n', 'gV', function() vim.api.nvim_feedkeys("`[" .. vim.fn.strpart(vim.fn.getregtype(), 0, 1) .. "`]", "n", false) end, { desc =  "Switch to VISUAL using last paste/change" })

-- vim.keymap.set("n", "gV", [['`[' . strpart(getregtype(), 0, 1) . '`]']], { expr = true })

-- TODO: control
vim.keymap.set('!',[[<M-\>]], [[<CMD>s/\(^.*\zs\(\s*\)\)\%#\s*/\=cursor(0,strlen(submatch(1))-strlen(submatch(2)))<CR>]])

-- vim.keymap.set('n', '<C-S-R>', "<CMD>exec 'undo' undotree()['seq_last']<CR>")

-- https://github.com/LazyVim/LazyVim/blob/13a4a84e3485a36e64055365665a45dc82b6bf71/lua/lazyvim/config/keymaps.lua#L64
-- Add undo break-points
vim.keymap.set("i", ",", ",<c-g>u")
vim.keymap.set("i", ".", ".<c-g>u")
vim.keymap.set("i", ";", ";<c-g>u")
vim.keymap.set("t", "<s-esc>", [[<c-\><c-n>]])
-- vim.keymap.set("n", "<s-esc>", [[]])

#+end_src

** File Operations
#+begin_src lua
-- INFO: "<C-R>=expand("%:t")<CR>" -> expands the current filename in the command line

-- from https://yazi-rs.github.io/docs/quick-start
-- what about g<c-y>? y and then modifier?
vim.keymap.set('n', [[<space>yy]], [[<cmd>let @" = expand("%:p")->fnamemodify(':~')   | echo 'cb> ' . @"<CR>]])
vim.keymap.set('n', [[<space>y.]], [[<cmd>let @" = expand("%:.")  | echo 'cb> ' . @"<CR>]])
-- d immediate dir and D complete?
vim.keymap.set('n', [[<space>yd]], [[<cmd>let @" = expand("%:p:h")->fnamemodify(':~') | echo 'cb> ' . @"<CR>]])
vim.keymap.set('n', [[<space>yf]], [[<cmd>let @" = expand("%:t") | echo 'cb> ' . @"<CR>]])
-- n stands for name? good if so
vim.keymap.set('n', [[<space>yn]], [[<cmd>let @" = expand("%:t:r") | echo 'cb> ' . @"<CR>]])
-- yank filename + line [souce :))]
vim.keymap.set('n', [[<space>yl]], [[<cmd>let @" = expand("%:p")->fnamemodify(':~') .. ":" .. line(".")   | echo 'cb> ' . @"<CR>]])

-- also create concatente yank
-- create also for normal mode
-- maybe also add last line?
vim.keymap.set('x','<space>y',function()
		  vim.cmd.norm('y')
        vim.fn.setreg(vim.v.register, vim.list_extend({vim.fn.fnamemodify(vim.fn.expand("%:p"),':~') .. ":" .. vim.fn.line("'<")},vim.fn.getreg(vim.v.register,'',true)))
end)


lua/config/options.lua
-- vim: foldmethod=expr foldexpr=getline(v\:lnum)=~'^--\ ┣'?'>1'\:getline(v\:lnum)=~'^--\ ─'?'>2'\:'=':
-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ TODO: ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- find nice configs/categories
-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ General ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- " ~~TODO: usa <space> come leader per programming, comma per altro~~
-- let mapleader = "\<Space>"
-- Set leader outside of keymaps to ensure that plugins know what to bind to
-- vim.g.mapleader = " "
-- look for the 2nd letter in the keymap to know which hand to use as leader
-- vim.g.mapleader = vim.keycode"<F13>"
-- vim.g.mapleader = vim.keycode"<f14>"
-- in accordanza con emacs...
vim.g.mapleader = vim.keycode"<space>"
-- vim.g.mapleader = "jk"
-- what about using mapleader for personal mappings e localleader for plugins? (or viceversa?) (https://www.reddit.com/r/neovim/comments/wta8k2/why_not_use_space_as_leader_key/)
-- for languages/filetype mappings?
-- f keys also useful so u can use them in insert/visual mode whichkey mappings
-- not too sure about this
-- vim.g.maplocalleader = vim.keycode[[<f15>]]
vim.g.maplocalleader = vim.keycode[[\]]
-- not sure
-- vim.g.maplocalleader = vim.keycode[[<c-c>]]

vim.opt.mouse:remove('a') -- Not use mouse
vim.opt.mousescroll  = 'ver:25,hor:6' -- Customize mouse scroll

vim.opt.showbreak="↪"

vim.opt.timeoutlen = 500 -- Mapping waiting time
-- Tip: per tradurre più parole usa visual mode
-- autocmd Filetype markdown,vimwiki set keywordprg=trans\ :it
-- autocmd Filetype markdown,vimwiki set keywordprg=trans\ en:it

-- https://github.com/neovim/neovim/issues/8427
-- per il momento usala, trova fix per :e, telescope commands, etc...
-- vim.opt.autochdir = true

-- aumenta fzf history
vim.opt.shada={"!","'500","<50","s10","h"}

vim.g.sessionoptions = "buffers,curdir,folds,help,tabpages,winsize,terminal" -- removed blank

-- exclude hidden buffers when saving session {for possession.nvim}
vim.opt.sessionoptions:remove('buffers')

vim.g.health = { style = 'float' }
-- vim.o.winborder='border'

vim.wo.signcolumn = "yes:1"
-- It messes up cpfd'json
-- vim.o.fileencodings = "utf-8,sjis,euc-jp,latin"
vim.o.title = true
vim.o.autoindent = true

vim.o.switchbuf    = 'usetab'       -- Use already opened buffers when switching

vim.opt.joinspaces = false

-- FIX: doesn't work with autopairs
-- vim.opt.digraph = true
-- fai che se ctrl-k non e precedura da ctrl-j allora digraph? maybe use snippet lua to check if first item is selected?

-- lualine
-- how to show full parent directory when the window is small?
-- vim.opt.showtabline = 1
-- autocmd FileType * set showtabline=0

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Spelling ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- vim.wo.spell = true
-- vim.o.spelllang    = 'en,uk,it,ru,ar'   -- Define spelling dictionaries
vim.o.spelloptions = 'camel'      -- Treat parts of camelCase words as seprate words
vim.opt.complete:append('kspell') -- Add spellcheck options for autocomplete
vim.opt.complete:remove('t')      -- Don't use tags for completion

vim.opt.dictionary = '/usr/share/dict/words'

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ File management ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- Not save backup
-- vim.opt.backup = false
-- vim.opt.swapfile = false
-- vim.opt.writebackup = false
-- Undotree
vim.opt.undolevels = 10000         -- use many levels of undo
-- vim.cmd([[
-- if has('persistent_undo')
--   set undodir=$HOME/.vim/undo
--   set undofile 
-- endif 
-- ]])
-- Persistent undo
-- vim.opt.undofile = true                -- Save undo's after file closes
-- vim.opt.undodir="$HOME/.vim/undo" -- where to save undo histories
-- vim.opt.undolevels = 1000         -- How many undos
-- vim.opt.undoreload = 10000        -- number of lines to save for undo
-- Accept modeline of each file
vim.opt.modeline = true
-- no idea why I can make this local...
vim.opt.modelineexpr = true
-- after eventual shebang and emacs...
vim.opt.modelines = 3
-- -- Encoding
-- opt.fileencoding = 'utf-8'
-- -- End of file setting
-- opt.fileformat = 'unix'



-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Text layout ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- -- Tabs, indent
vim.opt.expandtab = true       -- When this option is enabled, vi will use spaces instead of tabs
vim.opt.shiftwidth  = 2 -- Determines the amount of whitespace to add in normal mode
vim.opt.smartindent = true
vim.opt.tabstop     = 2 -- Width of tab character
vim.opt.softtabstop = 2 -- Fine tunes the amount of white space to be added

vim.opt.wrap = false
-- go to previous/next line with h,l,left arrow and right arrow
-- when cursor reaches end/beginning of line
-- opt.whichwrap:append "<>[]hl"
vim.opt.list = false
vim.opt.listchars = { trail = "-", tab = [[▸ ]], nbsp = "␣", eol = "$", extends = "❯", precedes ="❮" }
-- vim.opt.fillchars:append { diff = "╱" }
-- vim.opt_global.fillchars:append { diff = "╱" }

-- opt.fillchars = { eob = " " }

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Memory, CPU ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
vim.opt.hidden = true           -- Enable background buffers
vim.opt.history = 10000           -- Remember N lines in history
-- vim.opt.lazyredraw = true       -- Faster scrolling
vim.opt.synmaxcol = 240         -- Max column for syntax highlight
vim.opt.updatetime = 250        -- ms to wait for trigger an event

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Neovim UI ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- dunno if this is useful since I'm using leap.nvim
-- set relativenumber number
vim.opt.number = true           -- Show line number
-- vim.cmd[[let &stc='%=%{v:virtnum>0?"":printf("%X",v:lnum)} ']]
vim.opt.cursorline = true -- Highlight current line
vim.opt.showmatch = true        -- Highlight matching parenthesis
vim.opt.matchtime = 1
-- vim.opt.colorcolumn = '80'      -- Line lenght marker at 80 columns

-- like emacs
vim.opt.splitright = true       -- Vertical split to the right
vim.opt.splitbelow = true       -- Horizontal split to the bottom

vim.opt.ignorecase = true       -- Ignore case letters when search
vim.opt.smartcase = true        -- Ignore lowercase for the whole pattern
vim.opt.linebreak = true        -- Wrap on word boundary
vim.opt.termguicolors = true    -- Enable 24-bit RGB colors
-- hi Normal guibg=NONE ctermbg=NONE
-- vim.opt.laststatus = 2
vim.opt.laststatus=3            -- Set global statusline
-- set showtabline=2
vim.opt.showmode = false -- Not show redundant mode line with airline
-- Sexy Folds (https://www.reddit.com/r/neovim/comments/psl8rq/sexy_folds/)
vim.wo.foldenable = false
vim.wo.foldmethod = "expr"
vim.wo.foldexpr = "nvim_treesitter#foldexpr()"

-- https://www.reddit.com/r/neovim/comments/1h34lr4/neovim_now_has_the_builtin_lsp_folding_support/
-- vim.o.foldmethod = "expr"
-- vim.o.foldenable= false
-- fallback to treesitter if no lsp capabilities?
  -- vim.o.foldexpr = "v:lua.vim.lsp.foldexpr()"
-- vim.o.foldexpr = "nvim_treesitter#foldexpr()"

-- vim.wo.foldtext = [[substitute(getline(v:foldstart),'\\t',repeat('\ ',&tabstop),'g').'...'.trim(getline(v:foldend)) ]]
vim.wo.foldnestmax = 3
vim.wo.foldminlines = 1

-- vim.opt.foldopen = 'block,hor,mark,percent,quickfix,tag,jump,search,undo' -- What movements open folds


-- maybe create highlight.lua file?
vim.cmd('hi Visual guibg=peru guifg=Black')
-- hi Visual guibg=#000066 guifg=#00ff99


vim.opt.infercase = true                       -- Adjust completions to match case

-- vim.opt.more = false

-- silent! setglobal dictionary+=/usr/share/dict/words


vim.opt.mps:append('<:>')

-- stop auto-insertion of comment symbols.
vim.opt.formatoptions:remove{ "c", "r", "o" }
-- vim.o.formatoptions = 'rqnl1j' -- Improve comment editing

-- Define pattern for a start of 'numbered' list. This is responsible for
-- correct formatting of lists when using `gw`. This basically reads as 'at
-- least one special character (digit, -, +, *) possibly followed some
-- punctuation (. or `)`) followed by at least one space is a start of list
-- item'
vim.o.formatlistpat = [[^\s*[0-9\-\+\*]\+[\.\)]*\s\+]]

-- make the search recursive
vim.opt.path:append('**')

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Command mode ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- When completing, show all options, insert common prefix, then iterate
vim.opt.wildmode = {list = "longest", "full"}
vim.opt.wildignore = 'deps,.svn,CVS,.git,.hg,*.o,*.a,*.class,*.mo,*.la,*.so,*.obj,*.swp,*.jpg,*.png,*.xpm,*.gif,.DS_Store,*.aux,*.out,*.toc'
-- https://github.com/search?q=%2Fwildignore.*color%2F+language%3Alua+&type=code
--https://github.com/qw457812/dotfiles/blob/5d2090a1897806983945ad3d56ffc9db6c6039cc/dot_config/nvim/lua/config/options.lua#L17
vim.opt.wildignore:append(vim.api.nvim_get_runtime_file("colors/*.{vim,lua}", true))

-- Keep same column when going visual-block column
vim.opt.sol = false

-- Enable relative numbers only in Normal mode, and absolute numbers only in Insert mode
-- augroup toggle_relative_number
-- autocmd InsertEnter * :setlocal norelativenumber nonumber
-- autocmd InsertLeave * :setlocal relativenumber number

vim.opt.autoindent = true
vim.opt.smartindent = true
vim.opt.encoding = "utf-8"

-- vim.opt.diffopt:append({'vertical'})
-- [The linematch diffopt makes builtin diff so sweat! : r/neovim](https://www.reddit.com/r/neovim/comments/1ihpvaf/the_linematch_diffopt_makes_builtin_diff_so_sweat/)
vim.opt.diffopt="filler,internal,closeoff,algorithm:histogram,context:5,linematch:60"
-- set diffopt=vertical,filler,internal,algorithm:histogram,indent-heuristic

-- display completion matches on your status line
vim.opt.wildmenu = true

-- don't use included files and tag complete for better performances. Add this to your vimrc for these options :
-- set complete-=t
-- set complete-=i

vim.opt.inccommand="split"

-- hi Cursor guifg=green guibg=green
-- TODO: hide cursor after some time?
vim.cmd[[
"hi Cursor2 guifg=red guibg=Violet
" change color when inside colored text? like hex plugin?
" light green when dark background, rose when light background
hi Cursor2 guifg=red guibg=#00ffa0
"maybe use in insert mode?
set guicursor=n-v-c:block-Cursor/lCursor,i-n:Cursor2/lCursor2
]]

-- set guicursor=n-c-v-i:block-nCursor
-- set guicursor=i:blinkoff400-blinkon250


-- https://github.com/dettonijr/spass-syntax-highlight
-- au BufRead,BufNewFile *.dfg set filetype=dfg

-- set guifont=DroidSansMono\ Nerd\ Font\ 11

-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Theme ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
vim.o.background = 'dark'
-- vim.cmd('colorscheme Tomorrow-Night')
-- vim.cmd('colorscheme habamax')


-- ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫ Startup ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
-- Disable nvim intro
vim.opt.shortmess:append "sI"
-- -- Show shorten messages in UI
-- opt.shortmess:append('filmnrxoOtT')

vim.o.exrc=true
